File: Solutions/Neetcode150/BinarySearch/6-TimeBasedKeyValueStore/solution.ts
export class TimeMap {
  store: Map<string, [string, number][]>; // key: [value, timestamp]

  constructor() {
    this.store = new Map();
  }

  set(key: string, value: string, timestamp: number): void {
    const currValues = this.store.get(key);
    if (currValues) {
      currValues.push([value, timestamp]);
    } else {
      this.store.set(key, [[value, timestamp]]);
    }
  }

  get(key: string, timestamp: number): string {
    const currValues = this.store.get(key);
    if (!currValues) return "";

    let left = 0;
    let right = currValues.length - 1;
    let result = "";

    while (left <= right) {
      const mid = Math.floor((left + right) / 2);

      if (currValues[mid][1] <= timestamp) {
        result = currValues[mid][0];
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return result;
  }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * var obj = new TimeMap()
 * obj.set(key,value,timestamp)
 * var param_2 = obj.get(key,timestamp)
 */

File: Solutions/Neetcode150/BinarySearch/6-TimeBasedKeyValueStore/solution.py
from collections import defaultdict


class TimeMap:
    def __init__(self):
        self.map = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.map[key].append([value, timestamp])

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.map:
            return ""

        vals = self.map[key]
        res = ""

        l, r = 0, len(vals) - 1
        while l <= r:
            mid = (l + r) // 2
            if vals[mid][1] <= timestamp:
                res = vals[mid][0]
                l = mid + 1
            else:
                r = mid - 1

        return res


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
